# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [1]:
#1
import pandas as pd

url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"
df = pd.read_csv(url)

#dimensions of the dataset

print(f"The dataset has {df.shape[0]} rows and {df.shape[1]} columns.")
display(df.head())


The dataset has 4008 rows and 11 columns.


Customer          ST GENDER             Education Customer Lifetime Value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    Income  Monthly Premium Auto Number of Open Complaints     Policy Type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   Vehicle Class  Total Claim Amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323

In [2]:
#2
#determine the data types 

df_types = pd.DataFrame(df.dtypes).reset_index()
df_types.columns = ['column_name', 'type']
num_rows, num_columns = df.shape
print("Data types of each column:")
pd.DataFrame(df_types)

Data types of each column:


column_name     type
0                    Customer   object
1                          ST   object
2                      GENDER   object
3                   Education   object
4     Customer Lifetime Value   object
5                      Income  float64
6        Monthly Premium Auto  float64
7   Number of Open Complaints   object
8                 Policy Type   object
9               Vehicle Class   object
10         Total Claim Amount  float64

#Customer Lifetime Value       object
#Number of Open Complaints     object

'CLV' and 'Number of open complains' are numerical columns, since they contain numerical information. Therefore both columns should have data types like int or float, while categorical variables might have data types like object or category. 

However, before converting 'CLV' and 'Number of open complains' into numerical columns, they should be first checked for NaN values and inconsistencies (e.g if there are any letters).

In [3]:
# Convert columns to numerical type AFTER CLEANING
#columns_to_convert = ["Customer Lifetime Value", "Number of Open Complaints"]

#for column in columns_to_convert:
    #df[column] = pd.to_numeric(df[column], errors='coerce')

#print("\nUpdated Data Types after Conversion:")
#print(df.dtypes)

In [4]:
#3
# Display the number of unique values for each column
print("Number of unique values for each column:")
unic_num = pd.DataFrame(df.nunique())
display(unic_num)

# Identify categorical columns excluding the specified column(assumption: if unique values <= 10% of total rows, consider as categorical)
mistype_column = ["Number of Open Complaints", "Customer Lifetime Value"]  
categorical_columns = df.columns[(df.nunique() <= 10) & (df.columns != mistype_column[0])]
categorical_columns_table = pd.DataFrame(categorical_columns)
categorical_columns_table.columns = ['categorical_columns']
display(categorical_columns_table)


# Describe unique values for each categorical column
for column in categorical_columns:
    unique_values = df[column].unique()
    print(f"\nUnique values for {column}:\n{unique_values}")

Number of unique values for each column:


0
Customer                   1071
ST                            8
GENDER                        5
Education                     6
Customer Lifetime Value    1027
Income                      774
Monthly Premium Auto        132
Number of Open Complaints     6
Policy Type                   3
Vehicle Class                 6
Total Claim Amount          761

categorical_columns
0                  ST
1              GENDER
2           Education
3         Policy Type
4       Vehicle Class


Unique values for ST:
['Washington' 'Arizona' 'Nevada' 'California' 'Oregon' 'Cali' 'AZ' 'WA'
 nan]

Unique values for GENDER:
[nan 'F' 'M' 'Femal' 'Male' 'female']

Unique values for Education:
['Master' 'Bachelor' 'High School or Below' 'College' 'Bachelors' 'Doctor'
 nan]

Unique values for Policy Type:
['Personal Auto' 'Corporate Auto' 'Special Auto' nan]

Unique values for Vehicle Class:
['Four-Door Car' 'Two-Door Car' 'SUV' 'Luxury SUV' 'Sports Car'
 'Luxury Car' nan]


Some of types can also be transformed into 'categorical' type. For our analysis such columns can be determined by quantity of unique values less than 10. But although the column 'Number of Open Complaints' has unique values less than 10, it can not be transformed into categorical type due to the nature of its content (number of complains).  

As we can see in the description of categorical columns, there are inconsistencies in multiple columns. 

   For example, in 'ST' values 'California' and 'Cali' shold be merged in one value, as should 'Washington' and 'WA', 'Arizona' and 'AZ'. 

   In column 'GENDER' there are similar inconsistencies in values. They should be assigned to 3 categories 'M', 'F' and 'D'(diverse).

   I column 'Education' the values 'Bachelor' and 'Bachelors' should also be assighned to single value.

The next mistake which also appears in multiple columns is NaN. Which can eliminated as a next step of cleaning process. 

In [5]:
#display the range of values for numerical columns
numerical_columns = df.select_dtypes(include=["int64", "float64"]).columns

for column in numerical_columns:
    min_value = df[column].min()
    max_value = df[column].max()
    print(f" The range of column {column} is: {min_value} - {max_value}")


 The range of column Income is: 0.0 - 99960.0
 The range of column Monthly Premium Auto is: 61.0 - 35354.0
 The range of column Total Claim Amount is: 0.382107 - 2893.239678


Because the column 'Customer Lifetime Value' wasn't yet transforemed to numerical value, it's not automatically displayed among numerical columns.

The range of other 3 numarical columns shows the minimum and the maximum of a corresponding column.

In [7]:
#summary statistics fro numerical columns
stat_summary = df[numerical_columns].describe() 
stat_summary = stat_summary.round(2)
print("Summary Statistics for Numerical Columns:")
display(stat_summary)
    

Summary Statistics for Numerical Columns:


Income  Monthly Premium Auto  Total Claim Amount
count   1071.00               1071.00             1071.00
mean   39295.70                193.23              404.99
std    30469.43               1601.19              293.03
min        0.00                 61.00                0.38
25%    14072.00                 68.00              202.16
50%    36234.00                 83.00              354.73
75%    64631.00                109.50              532.80
max    99960.00              35354.00             2893.24

Income:

   Distribution: The distribution of income is right-skewed, as the mean (39,295.70) is greater than the median (36,234.00). This suggests that there are some higher-income outliers pulling the mean to the right.
   Variability: The standard deviation is relatively high (30,469.43), indicating a wide range of incomes.
                
Monthly Premium Auto:

   Premium Range: The monthly premium for auto insurance varies significantly, we can see it from the large standard deviation (1,601.19).
   Skewness: Similar to income, the distribution is likely right-skewed, with the mean (193.23), which is higher than the median (83.00).
    The presence of a significantly higher maximum premium (35,354.00) compared to the 75th percentile suggests the presence of outliers.
    
Total Claim Amount:

   Claim Distribution: The distribution of total claim amounts is right-skewed, as the mean (404.99) is greater than the median (354.73).
   Claim Range: The standard deviation (293.03) indicates some variability in claim amounts, but it is not as high as the variability in income or monthly premiums.
    Similar to premiums, the presence of a higher maximum claim amount (2,893.24) compared to the 75th percentile indicates potential outliers.

Recommendations:

Further exploration is needed to understand and potentially handle outliers in income, monthly premiums, and total claim amounts.
Consideration of any zero-income entries to determine whether they are valid or indicate missing data.
Visualizations, such as histograms or box plots, can provide additional insights into the distributions and identify potential outliers more easily.


In [71]:
#summary statistics fro categorical columns
categorical_summary = df[categorical_columns].dropna().describe(include='all')  

print("Summary Statistics for Categorical Columns:")
display(categorical_summary)


Summary Statistics for Categorical Columns:


ST GENDER Education    Policy Type  Vehicle Class
count      954    954       954            954            954
unique       8      5         6              3              6
top     Oregon      F  Bachelor  Personal Auto  Four-Door Car
freq       276    457       284            691            508

State (ST):
    The dataset includes entries from multiple states, with Oregon (OR) being the most represented state.
    Further exploration of regional patterns and potential state-specific trends could provide valuable insights.

Gender:
    The dataset is more heavily skewed toward females, with Female (F) being the most frequently occurring gender.
    It might be beneficial to investigate the reasons behind the gender imbalance and assess whether it has any implications for the analysis.

Education:
    Policies with Bachelor's degrees are the most common, and the education distribution appears somewhat balanced.
    However, as it was previously said, cleaning measures should be proveded for the further analysis.

Policy Type:
    Personal Auto policies dominate the dataset, indicating a significant focus on personal automobile insurance.
    
Vehicle Class:
    Four-Door Car is the most frequently occurring vehicle class.
    Understanding the popularity of certain vehicle classes may inform marketing strategies or product offerings.

General Recommendations:

   Data Quality Check: Explore the data for anomalies, missing data, duplicates and errors that might skew the results.
   Visualization: Utilize visualizations to enhance the understanding of categorical distributions and facilitate communication of insights.

## Challenge 2: analyzing the data

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [11]:
location = df['ST']
location_frequencies = location.value_counts()
top_5_less_common_location = location_frequencies.nsmallest(5)
print("Top 5 Less Common Customer Locations:")
display(top_5_less_common_location)

Top 5 Less Common Customer Locations:


AZ             25
WA             30
Washington     81
Nevada         98
Cali          120
Name: ST, dtype: int64

### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [16]:
policy_types = df['Policy Type']
policy_type_counts = policy_types.value_counts()
most_sold_policy = policy_type_counts.nlargest(1)
print("Total Number of Policies Sold for Each Policy Type:")
display(policy_type_counts)

print("\nPolicy Type with the Highest Number of Policies Sold:")
display(most_sold_policy)


Total Number of Policies Sold for Each Policy Type:


Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: Policy Type, dtype: int64


Policy Type with the Highest Number of Policies Sold:


Personal Auto    780
Name: Policy Type, dtype: int64

### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [22]:
policy_income_data = df[['Policy Type', 'Income']]

# Create two separate DataFrames for Personal Auto and Corporate Auto policies
personal_auto_df = policy_income_data.loc[policy_income_data['Policy Type'] == 'Personal Auto']
corporate_auto_df = policy_income_data.loc[policy_income_data['Policy Type'] == 'Corporate Auto']

# Calculate the average income for each policy type
average_income_personal_auto = round(personal_auto_df['Income'].mean(), 2)
average_income_corporate_auto = round(corporate_auto_df['Income'].mean(), 2)

# Display the results
print("Average Income for Personal Auto Policies:", average_income_personal_auto)
print("Average Income for Corporate Auto Policies:", average_income_corporate_auto)

if average_income_personal_auto > average_income_corporate_auto:
    print("Customers with Personal Auto have a higher average income.")
elif average_income_personal_auto < average_income_corporate_auto:
    print("Customers with Corporate Auto have a higher average income.")
else:
    print("Average incomes are the same for both policy types.")

Average Income for Personal Auto Policies: 38180.7
Average Income for Corporate Auto Policies: 41390.31
Customers with Corporate Auto have a higher average income.


### Bonus: Exercise 4


Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 

*Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.*

*Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.*

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

In [24]:
total_claim_amounts = df['Total Claim Amount']
claim_amount_stats = total_claim_amounts.describe().round(2)

#75th percentile
claim_amount_75th_percentile = total_claim_amounts.quantile(0.75).round(2)

#DataFrame with customers having a policy claim amount greater than the 75th percentile
high_claim_amount_customers = df[df['Total Claim Amount'] > claim_amount_75th_percentile]

#summary statistics about the high policy claim amount data
high_claim_amount_stats = high_claim_amount_customers['Total Claim Amount'].describe().round(2)

print("Statistics for Total Claim Amount:")
print(claim_amount_stats)

print("\n75th Percentile for Total Claim Amount:", claim_amount_75th_percentile)

print("\nSummary Statistics for High Policy Claim Amount Data:")
print(high_claim_amount_stats)

Statistics for Total Claim Amount:
count    1071.00
mean      404.99
std       293.03
min         0.38
25%       202.16
50%       354.73
75%       532.80
max      2893.24
Name: Total Claim Amount, dtype: float64

75th Percentile for Total Claim Amount: 532.8

Summary Statistics for High Policy Claim Amount Data:
count     264.00
mean      782.23
std       292.75
min       537.60
25%       606.52
50%       679.60
75%       851.40
max      2893.24
Name: Total Claim Amount, dtype: float64


Customers with policy claim amounts above the 75th percentile (532.80) are considered as having high claim amounts.
The average claim amount for this subset is significantly higher (782.23) compared to the overall average (404.99).
The variability in claim amounts, as indicated by the standard deviation, is relatively high for both the overall dataset and the subset of high claim amounts.
The distribution of high claim amounts is evident in the higher values of the 25th, 50th, and 75th percentiles for the subset.